### PR TITLE
Phase 2 R2: Phase 1 Retries with Correct 3h Timeout (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Round 1 alphonse experiments ran with the wrong timeout (30 min instead of 3h). The most promising Phase 1 retries need a proper re-run. Additionally, we include ideas from the closed PR analysis that showed strong potential but were epoch-starved.

## Instructions

**CRITICAL: Verify MAX_TIMEOUT=180.0 and MAX_EPOCHS=500 before running ANY experiment!**

All use the new noam baseline with --lr 1.5e-3.

### GPU 0: Pressure-only surface loss (Phase 1 PR #1116 — p_oodc improved)
Focus surface gradient signal entirely on pressure. In surf_loss computation, only use channel 2 (pressure): `surf_err = abs_err[:, :, 2:3]` instead of all 3 channels.
`CUDA_VISIBLE_DEVICES=0 python train.py --lr 1.5e-3 --wandb_name "alphonse/p2r2-ponly-surf" --wandb_group "phase2-r2-retries" --agent alphonse`

### GPU 1: Extended noise annealing (Phase 1 PR #1085 — needs longer schedule)
Start with strong noise (vel=0.02, p=0.01), anneal to near-zero over 150 epochs: `noise_progress = min(1.0, epoch / 150)`.
`CUDA_VISIBLE_DEVICES=1 python train.py --lr 1.5e-3 --wandb_name "alphonse/p2r2-noise-anneal" --wandb_group "phase2-r2-retries" --agent alphonse`

### GPU 2: n_hidden=224 + n_layers=2 (Phase 1 PRs #1069/#1089 — width between 192 and 256)
`CUDA_VISIBLE_DEVICES=2 python train.py --lr 1.5e-3 --wandb_name "alphonse/p2r2-h224" --wandb_group "phase2-r2-retries" --agent alphonse`

### GPU 3: Output head dropout p=0.05 (Phase 1 PR #1278 — classic technique, needs long training)
Add dropout=0.05 ONLY in the output MLP2 (not in attention or main MLP). This selectively regularizes the prediction head.
`CUDA_VISIBLE_DEVICES=3 python train.py --lr 1.5e-3 --wandb_name "alphonse/p2r2-head-dropout" --wandb_group "phase2-r2-retries" --agent alphonse`

### GPU 4: slice_num=56 (Phase 1 PR #1291 — finer routing, was epoch-starved)
Change slice_num from 48 to 56 in model_config. Finer spatial routing with more training time.
`CUDA_VISIBLE_DEVICES=4 python train.py --lr 1.5e-3 --wandb_name "alphonse/p2r2-slice56" --wandb_group "phase2-r2-retries" --agent alphonse`

### GPU 5: Relative position bias in slice attention (Phase 1 PR #1503 — Swin-style RPB)
Add learnable relative position bias to slice token self-attention. Create `self.rpb = nn.Parameter(torch.zeros(slice_num, slice_num))` and add to attention logits.
`CUDA_VISIBLE_DEVICES=5 python train.py --lr 1.5e-3 --wandb_name "alphonse/p2r2-rpb" --wandb_group "phase2-r2-retries" --agent alphonse`

### GPU 6: No tandem curriculum (Phase 1 PR #1693 — tied with baseline)
Remove the tandem curriculum entirely — train on all data from epoch 1. Set tandem_curriculum_epochs=0.
`CUDA_VISIBLE_DEVICES=6 python train.py --lr 1.5e-3 --tandem_curriculum_epochs 0 --wandb_name "alphonse/p2r2-no-curriculum" --wandb_group "phase2-r2-retries" --agent alphonse`

### GPU 7: Fourier PE bug fix (Phase 1 PR #1669 — padding fix, missed baseline by 0.65%)
Fix coordinate padding in Fourier PE computation. This was a genuine bug fix that narrowly missed baseline in 30 min.
`CUDA_VISIBLE_DEVICES=7 python train.py --lr 1.5e-3 --wandb_name "alphonse/p2r2-pe-fix" --wandb_group "phase2-r2-retries" --agent alphonse`

## Baseline
| Metric | Value |
|--------|-------|
| val/loss | 0.710 |
| p_in | 14.6 Pa |
| p_oodc | 10.0 Pa |
| p_tan | 35.6 Pa |
| p_re | 25.9 Pa |

---

## Results

All 8 experiments ran for the full 3-hour budget (~244–245 epochs each; h224 reached 206 due to larger model). None beat the noam baseline (0.710), but margins are small (0.7–2.9%).

### Summary table

| Experiment | val/loss | Δ baseline | p_in (Pa) | p_oodc (Pa) | p_tan (Pa) | p_re (Pa) | Peak Mem | W&B run |
|---|---|---|---|---|---|---|---|---|
| **Baseline (noam)** | **0.710** | — | 14.6 | 10.0 | 35.6 | 25.9 | — | — |
| GPU7: pe-fix | 0.7151 | +0.7% | 15.0 | 10.4 | 35.6 | 25.7 | 25.8 GB | 7y5j8p3l |
| GPU1: noise-anneal | 0.7155 | +0.8% | 15.3 | 10.3 | 35.4 | 25.9 | 26.6 GB | mc43egln |
| GPU5: RPB | 0.7170 | +1.0% | 14.9 | 10.3 | 35.9 | 25.8 | 26.0 GB | njhtm4l3 |
| GPU3: head-dropout | 0.7183 | +1.2% | 14.9 | 10.3 | 36.0 | 25.6 | 26.0 GB | p6psy0t7 |
| GPU0: ponly-surf | 0.7210 | +1.5% | 15.3 | 10.2 | 36.2 | 25.7 | 26.6 GB | dhh3l9h9 |
| GPU4: slice56 | 0.7242 | +2.0% | 14.9 | 10.5 | 36.6 | 25.6 | 27.0 GB | 5eslm7u6 |
| GPU2: h224 | 0.7279 | +2.5% | 15.0 | 10.4 | 36.4 | 26.0 | 30.5 GB | rsrvb2pt |
| GPU6: no-curriculum | 0.7306 | +2.9% | 15.1 | 11.0 | 36.7 | 25.7 | 26.2 GB | rn5jpu1p |

### What happened

**None of the 8 techniques improved on the noam baseline.** The noam baseline appears to have already found a strong operating point at lr=1.5e-3 with the Phase 2 schedule, and the marginal changes tested here didn't help.

Key observations:

- **pe-fix (GPU7)** came closest to baseline (0.7151, +0.7%). The padding-aware normalization is a genuine correctness fix — it prevents padded-zero coordinates from distorting the min/max, giving cleaner Fourier PE for smaller meshes. That it barely differs from no-fix suggests either (a) most samples don't have much padding variation, or (b) the model already learned to discount padded regions.

- **noise-anneal (GPU1)** was second best (0.7155, +0.8%). Starting with stronger noise (vel=0.02 vs 0.015) then annealing more aggressively over 150 epochs had essentially no effect vs the baseline's 60-epoch annealing. The model converges to similar loss regardless.

- **RPB (GPU5)** (0.7170, +1.0%) added slice-token relative position bias which is theoretically useful for capturing token-to-token spatial relationships, but with only 48 slices the bias matrix may not generalize well across different mesh configurations.

- **head-dropout (GPU3)** (0.7183, +1.2%) with p=0.05 provides mild regularization but with 3h training the model is already well-regularized by noise injection and the EMA. The dropout may actually slightly impair convergence.

- **ponly-surf (GPU0)** (0.7210, +1.5%) focusing the surface loss purely on pressure slightly hurt both val/loss and p_oodc. Removing velocity channels from the surface gradient signal hurt. The velocity channels on the surface provide useful training signal that anchors the model's understanding of no-slip boundary conditions.

- **slice56 (GPU4)** (0.7242, +2.0%) adding 8 more slices increased memory (+4%) but hurt performance. The default 48 slices appears to be a good operating point for this dataset; more slices may over-partition the mesh, making each slice's representation less stable.

- **h224 (GPU2)** (0.7279, +2.5%) is actually slightly worse than n_hidden=192 at the same budget. The larger model took fewer epochs (206 vs 244), suggesting it converged slower per-epoch, not faster overall. The width increase to 224 didn't help and cost more memory (30.5 vs 26 GB).

- **no-curriculum (GPU6)** (0.7306, +2.9%) performed worst, confirming the tandem curriculum is **important**. Removing it hurt val_ood_cond most (11.0 vs baseline 10.0 Pa) — the model got confused early by tandem-transfer samples and never fully recovered. The Phase 2 tandem curriculum should be kept at 20 epochs.

### Visualization note

The visualization section failed at the end of each run due to a `visualize()` signature mismatch (function now requires additional args `val_ds`, `stats`, `device`). All training metrics were logged to W&B normally; only the final plot generation failed. This is a pre-existing issue in the visualization utility, not related to the changes in this PR.

### Suggested follow-ups

- **pe-fix + noise-anneal combo**: The two closest-to-baseline variants (pe-fix +0.7%, noise-anneal +0.8%) might combine additively — worth a single clean run.
- **RPB with better initialization**: Zero init RPB didn't help, but could try small normal init (std=0.01) to give the model a warm start toward useful bias patterns.
- **keep tandem curriculum**: GPU6 confirms this is load-bearing. Any future experiments must keep `tandem_curriculum_epochs >= 20`.
- **head-dropout at lower rate (p=0.01–0.02)**: p=0.05 may be too aggressive given the existing noise regularization. A lower dropout rate might regularize without hurting convergence.
- **The noam baseline is the floor** — none of these simpler modifications beat it. Future gains likely require more substantial architectural changes or data augmentation strategies.

### CODE REVIEW

**Comment by tcapelle** [comment]
## Round 2 Results — Phase 1 Retries

All experiments complete. Results:

| GPU | Experiment | val/loss | p_in | p_oodc | p_tan | p_re |
|-----|-----------|---------|------|--------|-------|------|
| 0 | strong-noise | 0.724 | 15.1 | 10.4 | 36.9 | 25.5 |
| 1 | ponly-surf | 0.727 | 15.1 | 10.5 | 37.2 | 25.7 |
| 2 | surf-amp | 0.728 | 15.4 | 11.1 | 36.3 | 25.8 |
| 3 | h224 | 0.737 | 15.0 | 10.6 | 37.3 | 25.9 |
| 4 | iso-flops | 0.739 | 15.5 | 10.8 | 37.3 | 25.7 |
| 5 | 12fourier | 0.757 | 15.2 | 11.4 | 37.6 | 26.1 |
| 6 | sinusoidal-pe | 0.762 | 15.8 | 11.9 | 37.9 | 26.0 |

**Findings:** Phase 1 retries all beat Phase 1 baseline but none beat Round 1 best. Strong noise and pressure-only surface loss are the best retries. n_hidden=224 and iso-FLOPs don't justify the extra parameters vs n_hidden=192.
